File: py/src/aembit_edge/internal/client/token_state.py
"""Token lifecycle helpers."""

from __future__ import annotations

import json
import math
from dataclasses import dataclass
from datetime import datetime, timezone

from ...retry import RetryPolicy
from ..protocol.retry import (
    merge_retry_overrides,
    merge_retry_policy,
    serialize_effective_retry_policy,
)

DEFAULT_AUTH_EXPIRY_SKEW_MS = 60_000


@dataclass(slots=True, kw_only=True)
class CachedTokenState:
    """Cached auth token state for a client instance."""

    access_token: str
    expires_at_ms: int | None
    resource_set: str | None = None
    auth_cache_key: str | None = None


def resolve_auth_expiry_skew_ms(value: object) -> int:
    """Resolve configured token expiry skew in milliseconds."""

    if value is None:
        return DEFAULT_AUTH_EXPIRY_SKEW_MS

    if isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value):
        return DEFAULT_AUTH_EXPIRY_SKEW_MS

    if value < 0:
        return DEFAULT_AUTH_EXPIRY_SKEW_MS

    return int(value)


def is_token_valid(
    token_state: CachedTokenState | None,
    now_ms: int,
    skew_ms: int,
) -> bool:
    """Return true when cached token state is still valid for use."""

    if token_state is None:
        return False

    if token_state.expires_at_ms is None:
        return True

    return now_ms < token_state.expires_at_ms - skew_ms


def format_expires_at(expires_at_ms: int | None) -> str | None:
    """Format an optional expiry timestamp as an ISO 8601 string."""

    if expires_at_ms is None:
        return None

    return (
        datetime.fromtimestamp(expires_at_ms / 1000, tz=timezone.utc)
        .isoformat(timespec="milliseconds")
        .replace("+00:00", "Z")
    )


def resolve_effective_resource_set(
    default_resource_set: str | None,
    request_resource_set: str | None,
) -> str | None:
    """Resolve request-level Resource Set override against client default."""

    return request_resource_set if request_resource_set is not None else default_resource_set


def serialize_auth_single_flight_key(
    *,
    resource_set: str | None,
    auth_cache_key: str | None,
    retry_key: str,
) -> str:
    """Serialize the auth single-flight key used for de-duplication."""

    return json.dumps(
        [resource_set, auth_cache_key, retry_key],
        separators=(",", ":"),
    )


def serialize_effective_retry_policy_key(
    *,
    base_retry: RetryPolicy | None,
    request_retry: RetryPolicy | None,
) -> str:
    """Serialize the merged effective retry policy into a stable cache key."""

    merged_override = merge_retry_overrides(base_retry, request_retry)
    effective_retry = merge_retry_policy(merged_override)
    return serialize_effective_retry_policy(effective_retry)

File: ts/src/internal/client/token-state.ts
// Copyright 2024-present Aembit, Inc.
// SPDX-License-Identifier: Apache-2.0
import { mergeRetryPolicy } from "../protocol/retry.js";
import { mergeRetryOverrides } from "../shared/retry-utils.js";
import type { RetryPolicyOverride } from "../../types/retry.js";

const DEFAULT_AUTH_EXPIRY_SKEW_MS = 60_000;

export interface CachedTokenState {
  accessToken: string;
  expiresAtMs: number | null;
  resourceSet?: string;
  authCacheKey?: string;
}

interface AuthSingleFlightKey {
  resourceSet?: string;
  authCacheKey?: string;
  retryKey: string;
}

/**
 * Resolves configured token expiry skew in milliseconds.
 */
export function resolveAuthExpirySkewMs(value: number | undefined): number {
  if (typeof value === "number" && Number.isFinite(value) && value >= 0) {
    return Math.floor(value);
  }

  return DEFAULT_AUTH_EXPIRY_SKEW_MS;
}

/**
 * Returns true when cached token state is still valid for use.
 */
export function isTokenValid(
  tokenState: CachedTokenState | undefined,
  nowMs: number,
  skewMs: number
): tokenState is CachedTokenState {
  if (!tokenState) {
    return false;
  }

  if (tokenState.expiresAtMs === null) {
    return true;
  }

  return nowMs < tokenState.expiresAtMs - skewMs;
}

/**
 * Formats optional expiry timestamp as ISO string for public session output.
 */
export function formatExpiresAt(expiresAtMs: number | null): string | null {
  if (expiresAtMs === null) {
    return null;
  }

  return new Date(expiresAtMs).toISOString();
}

/**
 * Resolves request-level Resource Set override against client default.
 */
export function resolveEffectiveResourceSet(
  defaultResourceSet: string | undefined,
  requestResourceSet: string | undefined
): string | undefined {
  return requestResourceSet ?? defaultResourceSet;
}

/**
 * Serializes single-flight key for in-flight auth de-duplication.
 */
export function serializeAuthSingleFlightKey(key: AuthSingleFlightKey): string {
  return JSON.stringify([key.resourceSet ?? null, key.authCacheKey ?? null, key.retryKey]);
}

/**
 * Serializes the effective retry policy into a stable cache key.
 */
export function serializeEffectiveRetryPolicyKey(
  baseRetry: RetryPolicyOverride | undefined,
  requestRetry: RetryPolicyOverride | undefined
): string {
  const mergedOverride = mergeRetryOverrides(baseRetry, requestRetry);
  const effectiveRetry = mergeRetryPolicy(mergedOverride);

  return JSON.stringify({
    enabled: effectiveRetry.enabled,
    maxAttempts: effectiveRetry.maxAttempts,
    baseDelayMs: effectiveRetry.baseDelayMs,
    maxDelayMs: effectiveRetry.maxDelayMs,
    jitter: effectiveRetry.jitter,
    retryableStatusCodes: normalizeRetryableStatusCodes(
      effectiveRetry.retryableStatusCodes
    )
  });
}

function normalizeRetryableStatusCodes(
  statusCodes: number[] | undefined
): number[] | undefined {
  if (!statusCodes || statusCodes.length === 0) {
    return undefined;
  }

  return [...new Set(statusCodes)].sort((left, right) => left - right);
}
